lib/timeline/timeline_controller.dart

```diff
@@ -0,0 +1,154 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import 'dart:async';
+
+import 'package:vm_service_lib/vm_service_lib.dart' hide TimelineEvent;
+
+import '../globals.dart';
+import 'timeline_protocol.dart';
+
+/// This class must not have direct dependencies on dart:html.
+///
+/// This allows tests of the complicated logic in this class to run on the VM
+/// and will help simplify porting this code to work with Hummingbird.
+class TimelineController {
+  // Frame data.
+  static const int maxFrames = 120;
+  List<TimelineFrame> frames = <TimelineFrame>[];
+  final StreamController<TimelineFrame> _frameAddedController =
+      StreamController<TimelineFrame>.broadcast();
+  Stream<TimelineFrame> get onFrameAdded => _frameAddedController.stream;
+  final StreamController<Null> _framesClearedController =
+      StreamController<Null>.broadcast();
+  Stream<Null> get onFramesCleared => _framesClearedController.stream;
+
+  // Timeline data.
+  List<TimelineThreadEvent> dartEvents = <TimelineThreadEvent>[];
+  List<TimelineThreadEvent> gpuEvents = <TimelineThreadEvent>[];
+
+  TimelineData _timelineData;
+  TimelineData get timelineData => _timelineData;
+
+  bool get hasStarted => timelineData != null;
+
+  bool _paused = false;
+  bool get paused => _paused;
+
+  void pause() {
+    _paused = true;
+
+    dartEvents.clear();
+    gpuEvents.clear();
+  }
+
+  void resume() {
+    _paused = false;
+  }
+
+  Future<void> startTimeline() async {
+    await serviceManager.service
+        .setVMTimelineFlags(<String>['GC', 'Dart', 'Embedder']);
+    await serviceManager.service.clearVMTimeline();
+
+    final Response response = await serviceManager.service.getVMTimeline();
+    final List<dynamic> list = response.json['traceEvents'];
+    final List<Map<String, dynamic>> traceEvents =
+        list.cast<Map<String, dynamic>>();
+
+    final List<TimelineEvent> events = traceEvents
+        .map((Map<String, dynamic> event) => TimelineEvent(event))
+        .where((TimelineEvent event) {
+      return event.name == 'thread_name';
+    }).toList();
+
+    final TimelineData timelineData = TimelineData();
+
+    for (TimelineEvent event in events) {
+      final TimelineThread thread =
+          TimelineThread(timelineData, event.args['name'], event.threadId);
+      // TODO(kenzie): once we have gpu/cpu distinction data from engine, only
+      // add the threads that contain those events.
+      timelineData.addThread(thread);
+    }
+
+    timelineData.onTimelineThreadEvent.listen((TimelineThreadEvent event) {
+      processTimelineEvent(timelineData.getThread(event.threadId), event);
+    });
+
+    _timelineData = timelineData;
+  }
+
+  void processTimelineEvent(TimelineThread thread, TimelineThreadEvent event) {
+    if (thread == null) {
+      return;
+    }
+
+    // io.flutter.1.ui, io.flutter.1.gpu
```

Comment:
Add a TODO warning just how fragile it is to use these names and noting we can't run tests with them.

Reply:
done.